### PR TITLE
RavenDB-6602 implementation

### PR DESCRIPTION
**Pull request by myarichuk**
there is still ReplicationWithFailover::LoadDocumentsWithFailOver() test to fix -> related to this issue

### CODE REVIEW

File: src/Raven.Client/Documents/DocumentStore.cs [hunk 1]
```diff
@@ -184,13 +181,42 @@ public override IDocumentSession OpenSession(SessionOptions options)
...
 
+        public async Task ForceUpdateTopologyFor(string databaseName = null)
+        {
+            var requestExecutor = GetRequestExecuter(databaseName);
+            await requestExecutor.UpdateTopologyAsync();
+        }
+
+        internal override void RedirectRequestExecuterTo(string url, string databaseName = null)
+        {
+            var lazy = _requestExecuters.AddOrUpdate(databaseName,
```
Comment (ayende) [inline src/Raven.Client/Documents/DocumentStore.cs:192]: Why is this called for? Doesn't seem to be doing anything

File: src/Raven.Client/Documents/DocumentStore.cs [hunk 2]
```diff
@@ -184,13 +181,42 @@ public override IDocumentSession OpenSession(SessionOptions options)
...
+        {
+            var lazy = _requestExecuters.AddOrUpdate(databaseName, 
+                newKey =>
+                {
+                    var newRequestExecutor = RequestExecutor.Create(url, newKey, ApiKey);
+                    newRequestExecutor.TopologyUpdate += () => OnTopologyUpdatedInternal(newKey);
+                    return newRequestExecutor;
+                }, 
+                (existingKey, existingLazy) =>
+                {
```
Comment (ayende) [inline src/Raven.Client/Documents/DocumentStore.cs:200]: Doesn't the existing value need to be disposed?

File: src/Raven.Client/Http/ReadBehavior.cs [hunk 3]
```diff
@@ -2,9 +2,9 @@
 {
     public enum ReadBehavior
     {
-        LeaderOnly,
-        LeaderWithFailover,
-        LeaderWithFailoverWhenRequestTimeSlaThresholdIsReached,
+        ConversationNodeOnly,
```
Comment (ayende) [inline src/Raven.Client/Http/ReadBehavior.cs:5]: What is `Conversation` ? That is not a term I'm familiar with.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -44,58 +48,75 @@ public class AggresiveCacheOptions
...
 
             if (requiresTopologyUpdates == false)
                 return;
 
             _updateTopologyTimer = new Timer(UpdateTopologyCallback, null, 0, Timeout.Infinite);
-            _updateFailingNodesStatus = new Timer(UpdateFailingNodesStatusCallback, null, TimeSpan.FromMinutes(1),
-                TimeSpan.FromMinutes(1));
+
+            //this would be handled by heartbeats between cluster nodes (that are not yet implemented)
+            //_updateFailingNodesStatus = new Timer(UpdateFailingNodesStatusCallback, null, TimeSpan.FromMinutes(1),
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:99]: This is not part of the client. The client merely update its topology every minutes, nothing more.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 5]
```diff
@@ -108,15 +129,17 @@ public static RequestExecutor CreateForSingleNode(string url, string databaseNam
...
 
-        public async Task<bool> UpdateTopology()
+        public async Task<bool> UpdateTopologyAsync()
         {
             if (_disposed)
                 return false;
-            bool lookTaken = false;
-            Monitor.TryEnter(this, 0, ref lookTaken);
+            
+            var lockTaken = _updateTopologySemaphore.Wait(0);
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:140]: What is actually being protected in this lock?

File: src/Raven.Client/Documents/DocumentStoreBase.cs [hunk 6]
```diff
@@ -197,6 +201,8 @@ protected virtual void AfterSessionCreated(InMemoryDocumentSessionOperations ses
 
         public abstract RequestExecutor GetRequestExecuter(string databaseName = null);
 
+        internal abstract void RedirectRequestExecuterTo(string url, string databaseName = null);
```
Comment (ayende) [inline src/Raven.Client/Documents/DocumentStoreBase.cs:204]: This is never called

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 7]
```diff
@@ -126,22 +149,52 @@ protected virtual void UpdateTopologyCallback(object _)
...
+                            }
+                            catch (Exception)
+                            {
+                                _nodeSelector.OnFailedRequest();
+                                continue;
+                            }
+
+                            //because we failed to contact a server, force update topology because it is likely that the etag 
+                            //hadn't changed, since the topology itself didn't change, it's a node that went down
+                            if (command.Result.Nodes.Count > 0 && _nodeSelector.OnUpdateTopology(command.Result, forceUpdate:true))
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:188]: why are you calling the topology in the middle of error handling?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 8]
```diff
@@ -195,15 +228,20 @@ public void Execute<TResult>(RavenCommand<TResult> command, JsonOperationContext
 
         public async Task ExecuteAsync<TResult>(RavenCommand<TResult> command, JsonOperationContext context, CancellationToken token = default(CancellationToken))
         {
-            var choosenNode = ChooseNodeForRequest(command);
+            //run it only the first time we execute this function
+            if (_requiresTopologyUpdates && Interlocked.CompareExchange(ref _hasUpdatedTopologyAtLeastOnce, 1, 0) == 0)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:232]: Bad, if we have concurrent calls, the first one will run this, the other will not wait

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 9]
```diff
@@ -238,29 +276,22 @@ public void Execute<TResult>(RavenCommand<TResult> command, JsonOperationContext
                     else
                     {
                         response = await command.SendAsync(client, request, token).ConfigureAwait(false);
+                        if (response.Headers.Contains(Constants.Headers.ForceTopologyUpdate))
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:279]: Under what conditions will the server send it?

Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs:279]: And why are we forcing the client to wait for the next topology when we already have the response for what it asked for?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 10]
```diff
@@ -302,15 +333,15 @@ private static HttpRequestMessage CreateRequest<TResult>(ServerNode node, RavenC
 
             request.RequestUri = new Uri(url);
 
-            if (node.CurrentToken != null)
-                request.Headers.Add("Raven-Authorization", node.CurrentToken);
+            if (node.ClusterToken != null)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:336]: This is not a cluster token, it is the current node _for that specific machine_

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Http/ReadBehavior.cs (hunk 3 above)
Comment (myarichuk) [reply src/Raven.Client/Http/ReadBehavior.cs:5]: I just transcribed it from our discussion about this. I will change this to "CurrentNodeOnly" kind of naming

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 9 above)
Comment (myarichuk) [reply src/Raven.Client/Http/RequestExecutor.cs:279]: Under what conditions will the server send it? --> we have talked that the server will send it when senator is changed or when there are any changes to topology that come from the server-side.
(currently I think that senator related stuff is not yet implemented)

**Pull request closed without merging**